src/util/bip341.rs

```diff
@@ -0,0 +1,556 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::IoError => write!(f, "IoError"),
+            Error::PrevoutsSize => write!(f, "Number of supplied prevouts differs from the number of inputs in transaction"),
+            Error::PrevoutIndex => write!(f, "The index requested is greater than available prevouts or different from the provided [Provided::Anyone] index"),
+            Error::PrevoutKind => write!(f, "A single prevout has been provided but all prevouts are needed without `ANYONECANPAY`")
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+impl Prevouts {
+    fn get_all(&self) -> Result<&[TxOut], Error> {
+        match self {
+            Prevouts::All(ref prevouts) => Ok(prevouts),
+            _ => Err(Error::PrevoutKind),
+        }
+    }
+
+    fn get(&self, input_index: usize) -> Result<&TxOut, Error> {
+        match self {
+            Prevouts::Anyone(index, prevout) => {
+                if input_index == *index {
+                    Ok(prevout)
+                } else {
+                    Err(Error::PrevoutIndex)
+                }
+            }
+            Prevouts::All(prevouts) => prevouts.get(input_index).ok_or(Error::PrevoutIndex),
+        }
+    }
+}
+
+/// A replacement for SigHashComponents which supports all sighash modes
+pub struct SigHashCache<'a> {
+    /// Access to transaction required for various introspection
+    tx: &'a Transaction,
+
+    /// Hash of all the previous outputs, computed as required
+    hash_prevouts: Option<sha256::Hash>,
+    /// Hash of all the input sequence nos, computed as required
+    hash_sequence: Option<sha256::Hash>,
+    /// Hash of all the outputs in this transaction, computed as required
+    hash_outputs: Option<sha256::Hash>,
+
+    /// Hash of all the prevout amounts, computed as required
+    hash_amounts: Option<sha256::Hash>,
+    /// Hash of all the prevout scriptpubkeys, computed as required
+    hash_scriptpubkeys: Option<sha256::Hash>,
+}
+
+impl<'a> SigHashCache<'a> {
+    /// Compute the sighash components from an unsigned transaction and auxiliary
+    /// in a lazy manner when required.
+    /// For the generated sighashes to be valid, no fields in the transaction may change except for
+    /// script_sig and witnesses.
+    pub fn new(tx: &'a Transaction) -> Self {
+        SigHashCache {
+            tx,
+            hash_prevouts: None,
+            hash_sequence: None,
+            hash_outputs: None,
+            hash_amounts: None,
+            hash_scriptpubkeys: None,
+        }
+    }
+
+    /// Encode the BIP341 signing data for any flag type into a given object implementing a
+    /// std::io::Write trait.
+    pub fn encode_signing_data_to<Write: io::Write>(
+        &mut self,
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
+        annex: Option<Vec<u8>>,
+        script_path: Option<ScriptPath>,
+        sighash_type: SigHashType,
```

Comment:
From
https://github.com/rust-bitcoin/rust-bitcoin/pull/628/commits/a783ceab229174b39a7b5d6cea8550558309c78d#r667735143

Should we have input as `u8` for `SigHashType`? Part of me thinks that extending SigHashType with another enum for `SigHashType::Default` might be super breaking downstream change. 

I am not super thrilled about a new `SighashTypeTaproot`, but that might be best way forward.

Reply:
I think a new `SighashTypeTaproot` with the same variant as `SigHashType` plus `SighashTypeTaproot::Default` is the best way, yes.

Maybe the name is not crystal-clear since it is introduced with taproot but will be usable also for pre-taproot sighashes

Reply:
Yup agreed, +1 for `SighashTypeTaproot`(or any other better name)

Reply:
it might make sense to just enum it because sighash types can diverge in the future (e.g., SighashTypeAnyprevout